File: python/gmetric.py
#!/usr/bin/env python

# This is the MIT License
# http://www.opensource.org/licenses/mit-license.php
#
# Copyright (c) 2007,2008 Nick Galbreath
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#

#
# Version 1.0 - 21-Apr-2007
#   initial
# Version 2.0 - 16-Nov-2008
#   made class Gmetric thread safe
#   made gmetrix xdr writers _and readers_
#   Now this only works for gmond 2.X packets, not tested with 3.X
#

from xdrlib import Packer, Unpacker
import socket

slope_str2int = {'zero':0,
                 'positive':1,
                 'negative':2,
                 'both':3,
                 'unspecified':4}

# could be autogenerated from previous but whatever
slope_int2str = {0: 'zero',
                 1: 'positive',
                 2: 'negative',
                 3: 'both',
                 4: 'unspecified'}

def gmetric_write(NAME, VAL, TYPE, UNITS, SLOPE, TMAX, DMAX):
    """
    Arguments are in all upper-case to match XML
    """
    packer = Packer()
    packer.pack_int(0)   # type gmetric
    packer.pack_string(TYPE)
    packer.pack_string(NAME)
    packer.pack_string(str(VAL))
    packer.pack_string(UNITS)
    packer.pack_int(slope_str2int[SLOPE]) # map slope string to int
    packer.pack_uint(int(TMAX))
    packer.pack_uint(int(DMAX))
    return packer.get_buffer()

def gmetric_read(msg):
    unpacker = Unpacker(msg)
    values = dict()
    unpacker.unpack_int()
    values['TYPE'] = unpacker.unpack_string()
    values['NAME'] = unpacker.unpack_string()
    values['VAL'] = unpacker.unpack_string()
    values['UNITS'] = unpacker.unpack_string()
    values['SLOPE'] = slope_int2str[unpacker.unpack_int()]
    values['TMAX'] = unpacker.unpack_uint()
    values['DMAX'] = unpacker.unpack_uint()
    unpacker.done()
    return values


class Gmetric:
    """
    Class to send gmetric/gmond 2.X packets

    Thread safe
    """

    type = ('', 'string', 'uint16', 'int16', 'uint32', 'int32', 'float',
            'double', 'timestamp')
    protocol = ('udp', 'multicast')

    def __init__(self, host, port, protocol):
        if protocol not in self.protocol:
            raise ValueError("Protocol must be one of: " + str(self.protocol))

        self.socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        if protocol == 'multicast':
            self.socket.setsockopt(socket.IPPROTO_IP,
                                   socket.IP_MULTICAST_TTL, 20)
        self.hostport = (host, int(port))
        #self.socket.connect(self.hostport)

    def send(self, NAME, VAL, TYPE='', UNITS='', SLOPE='both',
             TMAX=60, DMAX=0):
        if slopestr not in slope_str2int:
            raise ValueError("Slope must be one of: " + str(self.slope.keys()))
        if typestr not in self.type:
            raise ValueError("Type must be one of: " + str(self.type))
        if len(name) == 0:
            raise ValueError("Name must be non-empty")

        msg = gmetric_write(NAME, VAL, TYPE, UNITS, SLOPE, TMAX, DMAX)
        return self.socket.sendto(msg, self.hostport)


if __name__ == '__main__':
    import optparse
    parser = optparse.OptionParser()
    parser.add_option("", "--protocol", dest="protocol", default="udp",
                      help="The gmetric internet protocol, either udp or multicast, default udp")
    parser.add_option("", "--host",  dest="host",  default="127.0.0.1",
                      help="The gmond host to recieve the data")
    parser.add_option("", "--port",  dest="port",  default="8649",
                      help="The gmond port to recieve the data")
    parser.add_option("", "--name",  dest="name",  default="",
                      help="The name of the metric")
    parser.add_option("", "--value", dest="value", default="",
                      help="The value of the metric")
    parser.add_option("", "--units", dest="units", default="",
                      help="The units for the value, e.g. 'kb/sec'")
    parser.add_option("", "--slope", dest="slope", default="both",
                      help="The sign of the derivative of the value over time, one of zero, positive, negative, both, default both")
    parser.add_option("", "--type",  dest="type",  default="",
                      help="The value data type, one of string, int8, uint8, int16, uint16, int32, uint32, float, double")
    parser.add_option("", "--tmax",  dest="tmax",  default="60",
                      help="The maximum time in seconds between gmetric calls, default 60")
    parser.add_option("", "--dmax",  dest="dmax",  default="0",
                      help="The lifetime in seconds of this metric, default=0, meaning unlimited")
    (options,args) = parser.parse_args()

    g = Gmetric(options.host, options.port, options.protocol)
    g.send(options.name, options.value, options.type, options.units,
           options.slope, options.tmax, options.dmax)

File: gmetric2/java/gmetric.java


/*
 * This is the MIT License
 * http://www.opensource.org/licenses/mit-license.php
 *
 * Copyright (c) 2009 Nick Galbreath
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

import java.io.DataOutputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;

/**
 *
 *
 *
 */
class gmetric {

    public final static int SLOPE_ZERO         = 0;
    public final static int SLOPE_POSITIVE     = 1;
    public final static int SLOPE_NEGATIVE     = 2;
    public final static int SLOPE_BOTH         = 3;
    public final static int SLOPE_UNSPECIFIED  = 4;

    public final static String VALUE_STRING          = "string";
    public final static String VALUE_UNSIGNED_SHORT  = "uint16";
    public final static String VALUE_SHORT           = "int16";
    public final static String VALUE_UNSIGNED_INT    = "uint32";
    public final static String VALUE_INT             = "int32";
    public final static String VALUE_FLOAT           = "float";
    public final static String VALUE_DOUBLE          = "double";

    public static void send(InetAddress address, int port,
			    String name, String value, String type,
			    String units, int slope, int tmax, int dmax)
    {
	try {
	    DatagramSocket socket = new DatagramSocket();
	    byte[] buf = write(name, value, type, units, slope, tmax, dmax);
	    DatagramPacket p = new DatagramPacket(buf, buf.length, address, port);
	    socket.send(p);
	} catch (IOException e) {
	    // who cares
	}
    }

    public static void send(InetAddress address, int port,
			    String name, double dvalue,
			    String units, int slope, int tmax, int dmax)
    {
	try {
	    String value = Double.toString(dvalue);
	    DatagramSocket socket = new DatagramSocket();
	    byte[] buf = write(name, value, VALUE_DOUBLE, units, slope, tmax, dmax);
	    DatagramPacket p = new DatagramPacket(buf, buf.length, address, port);
	    socket.send(p);
	} catch (IOException e) {
	    // who cares
	}
    }

    public static void send(InetAddress address, int port,
			    String name, int dvalue,
			    String units, int slope, int tmax, int dmax)
    {
	try {
	    String value = Integer.toString(dvalue);
	    DatagramSocket socket = new DatagramSocket();
	    byte[] buf = write(name, value, VALUE_INT, units, slope, tmax, dmax);
	    DatagramPacket p = new DatagramPacket(buf, buf.length, address, port);
	    socket.send(p);
	} catch (IOException e) {
	    // who cares
	}
    }



    /*
     * EVERYTHING BELOW HERE YOU DON"T NEED TO USE
     */

    public static byte[] write(String name, String value, String type,
			       String units, int slope, int tmax, int dmax)
    {
	try {
	    ByteArrayOutputStream baos = new ByteArrayOutputStream();
	    DataOutputStream dos = new DataOutputStream(baos);
	    dos.writeInt(0);
	    writeXDRString(dos, type);
	    writeXDRString(dos, name);
	    writeXDRString(dos, value);
	    writeXDRString(dos, units);
	    dos.writeInt(slope);
	    dos.writeInt(tmax);
	    dos.writeInt(dmax);
	    return baos.toByteArray();
	} catch (IOException e) {
	    // really this is impossible
	    return null;
	}
    }

    private static void writeXDRString(DataOutputStream dos, String s) throws IOException
    {
	dos.writeInt(s.length());
	dos.writeBytes(s);
	int offset = s.length() % 4;
	if (offset != 0) {
	    for (int i = offset; i < 4; ++i) {
		dos.writeByte(0);
	    }
	}
	/*
	bytes[] b = s.getBytes("utf8");
	int len = b.length();
	dos.writeInt(len);
	dos.write(b, 0, len);
	*/
    }

    /*
     * Everything below here is just for testing
     */
    private static final byte[] HEXCHARS = {
	(byte)'0', (byte)'1', (byte)'2', (byte)'3',
	(byte)'4', (byte)'5', (byte)'6', (byte)'7',
	(byte)'8', (byte)'9', (byte)'a', (byte)'b',
	(byte)'c', (byte)'d', (byte)'e', (byte)'f'
    };

    private static String bytes2hex(byte[] raw)
    {
	try {
	    int pos = 0;
	    byte[] hex = new byte[2 * raw.length];
	    for (int i = 0; i < raw.length; ++i) {
		int v = raw[i] & 0xFF;
		hex[pos++] = HEXCHARS[v >>> 4];
		hex[pos++] = HEXCHARS[v & 0xF];
	    }

	    return new String(hex, "ASCII");
	} catch (UnsupportedEncodingException e) {
	    // impossible
	    return "";
	}
    }

    public static void selfTest()
    {
	String expected = "0000000000000006737472696e67000000000003666f6f00000000036261720000000000000000030000003c00000000";
	byte[] data = write("foo", "bar", VALUE_STRING, "", SLOPE_BOTH, 60, 0);
	String result = bytes2hex(data);
	if (!expected.equals(result)) {
	    System.out.println(expected);
	    System.out.println(result);
	    throw new RuntimeException("bad");
	}
    }

    public static void main(String args[]) throws Exception {
	selfTest();
    }
}
